### PR TITLE
The Prisoners puzzle

### PR DESCRIPTION
**Pull request by konnov**
This is a simple translation of the TLA+ puzzle. It's not working in REPL yet, as #236 is yet to be implemented (my next task).

- [x] Entries added to the respective `CHANGELOG.md` for any new functionality

### CODE REVIEW

**Review by shonfeder — APPROVED** [review]
Looks good!

A few minor questions and comments.

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 - command `test` to run unit tests (#634)
+- the [Prisoners puzzle](./examples/puzzles/prisoners) (#634)
```
[the remark this answers is not in the document]
Comment (konnov) [reply CHANGELOG.md:13]: I thought that the point of the changelog is to inform the users about new features and potentially breaking changes. In this sense, one more example is similar to a new feature or a piece of documentation.

File: examples/puzzles/prisoners/prisoners.qnt [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+ * https://github.com/tlaplus/Examples/blob/master/specifications/Prisoners/Prisoners.tla
+ *
+ * This specification is derived from the above TLA+ specification.
+ */
+module prisoners {
+    // The set of all prisoners
+    const allPrisoners: Set[str]
+
+    // The prisoner who is counting
+    var counter: str
```
[the remark this answers is not in the document]
Comment (konnov) [reply examples/puzzles/prisoners/prisoners.qnt:14]: There is not good reason actually. I was just surprised that it was a constant, but the text was saying that it was chosen arbitrarily. Let me make it a constant, so it does not deviate from the original spec.

File: examples/puzzles/prisoners/prisoners.qnt [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    // the state initializer
+    action init = all {
+        count' = 0,
+        // pick one prisoner as a counter
+        nondet c = allPrisoners.oneOf()
+        all {
+            counter' = c,
+            timesSwitched' = nonCounters.mapBy(p => 0),
+        },
```
[the remark this answers is not in the document]
Comment (konnov) [reply examples/puzzles/prisoners/prisoners.qnt:41]: oh, there is no good reason. Probably, an artefact of earlier code. Will fix.

File: examples/puzzles/prisoners/prisoners.qnt [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    */
+
+    /**
+      * The safety condition as a state invariant: Done true implies
+      * that every prisoner other than the counter has flipped switch A at
+      * least once--and hence has been in the room at least once. Since the
+      * counter increments the count only when in the room, and Done implies
+      * count > 0, it also implies that the counter has been in the room.
+      */
+    val safetyInv = nonCounters.forall(p => timesSwitched.get(p) > 0)
```
[the remark this answers is not in the document]
Comment (konnov) [reply examples/puzzles/prisoners/prisoners.qnt:114]: This invariant probably makes sense too, but this is not what is written in the original spec: https://github.com/tlaplus/Examples/blob/be442974429de555e4c8df0331b9de7dfdc25310/specifications/Prisoners/Prisoners.tla#L186-L205

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply CHANGELOG.md:13]: Yes, that’s what I take the purpose to be, but I don’t take documentation changes to be new features.

If we want to include this kind of thing in the changelog, maybe we should add a “documentation” section? WDYT?

File: examples/puzzles/prisoners/prisoners.qnt (hunk 4 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply examples/puzzles/prisoners/prisoners.qnt:111]: Aren’t we taking about https://github.com/tlaplus/Examples/blob/be442974429de555e4c8df0331b9de7dfdc25310/specifications/Prisoners/Prisoners.tla#L162 here?

[the remark this answers is not in the document]
Comment (konnov) [reply examples/puzzles/prisoners/prisoners.qnt:111]: Argh. Sorry, you are right :)

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (konnov) [reply CHANGELOG.md:13]: This makes sense to me!

**Pull request merged**